src/main/scala/is/hail/expr/ir/IR.scala

```diff
@@ -0,0 +1,74 @@
+package is.hail.expr.ir
+
+import is.hail.asm4s.TypeInfo
+
+import scala.language.existentials
+
+sealed case class PrimitiveOp(x: Long)
+
+sealed trait IR
+sealed case class NA(typ: PointedTypeInfo[_]) extends IR
+sealed case class I32(x: Int) extends IR
+sealed case class I64(x: Long) extends IR
+sealed case class F32(x: Float) extends IR
+sealed case class F64(x: Double) extends IR
+sealed case class True() extends IR
+sealed case class False() extends IR
+sealed case class If(cond: IR, cnsq: IR, altr: IR) extends IR
+sealed case class Let(name: String, value: IR, typ: TypeInfo[_], body: IR) extends IR
+sealed case class ApplyPrimitive(op: PrimitiveOp, args: Array[IR]) extends IR
+sealed case class LazyApplyPrimitive(op: PrimitiveOp, args: Array[IR]) extends IR
+sealed case class Lambda(name: String, body: IR) extends IR
+sealed case class Ref(name: String) extends IR
+sealed case class MakeArray(args: Array[IR], typ: TypeInfo[_]) extends IR
+sealed case class ArrayRef(a: IR, i: IR, typ: TypeInfo[_]) extends IR
+sealed case class MakeStruct(fields: Array[(String, TypeInfo[_], IR)]) extends IR
+sealed case class GetField(o: IR, name: String) extends IR
+sealed case class MapNA(name: String, value: IR, valueTyp: TypeInfo[_], body: IR, bodyTyp: PointedTypeInfo[_]) extends IR
+sealed case class In(i: Int) extends IR
+sealed case class Out(values: Array[IR]) extends IR
```

Comment:
We should think a bit more about I/O.  We should think about the larger structure instead of making something plugin-in to the existing evaluator.  Namely, I want the entire run-time computation graph represented in the AST for optimization before we run anything.  Therefore, something like annotate_variants_expr will need AST nodes that represent inserting into the `va` field and you will just return a long (a offset/pointer into the current Region).  But filter_variants will require a function that returns a boolean.  Although not necessarily missing, the generated code can handling missing.

Say no to boxing always, anywhere, ever.

Reply:
Do you envision storing all intermediate values in a region? It seems sufficient to use the JVM stack for these values, with explicit copying into/out-of the region.

That would demand adding `Move(destination: IR, source: IR)` to the IR which takes an intermediate value and places it at the location given by a particular `RegionValue`. `annotate_variants_expr` would be fully represented in the IR (I imagine the `RegionValue` corresponding to a destination annotation path would be passed as an argument to the compiled IR) . We probably also want `Read(rv: IR)` which reads a `RegionValue` as an intermediate.

`Read` is more expensive for `Array` or `Struct` than it is for `Int`.

It seems the IR needs a full suite of `RegionValue` manipulation operations. I don't want to unnecessarily round-trip through Read/Write. For example, to set the first element of an array (given to me as my first input) to forty-two:

```
(Move (RegionValueArrayAtOffset (In 0) 0) 42)
```
The annoying thing seems to be that `RegionValue` is not a `long`, it has a `Region` associated with it. I guess, I need to represent this explicitly:

```
(Move (RegionValue (In 0) (ApplyPrimitive + (In 1) (I32 0))) 42)
```
I feel like I'm getting deep into the weeds now. Tracking the flow of values through the annotation fields now reduces to pointer tracking.

Reply:
> Do you envision storing all intermediate values in a region? 

By what you seem mean by all, absolutely not.  Scalars (with missingness) should always be stored on the stack.  The question is how to handle aggregate objects (strings/binary, structs, arrays).

 > which takes an intermediate value and places it at the location given by a particular RegionValue.

I don't think it makes sense to place a value at the location of a RegionValue.  It makes sense to place a value in a region which gives a region value back.

There are three natural places to put aggregate values:
 - in the region being processed,
 - in a temporary region for the evaluation of this expression,
 - on the java heap.

The latter two are roughly the same, but I think the region will ultimately be more efficient (better be).  We already have code to implement either (RVB.addAnnotation, RVB.addRegionValue).

This goes back to our conversation on "physical types".  A physical type could represent a region value or something in the Java heap.

All that said, the simplest thing is probably to have a single fixed region associated with the evaluation and assume all aggregate values are stored there.

I don't think region values should be mutable.

 > It seems the IR needs a full suite of RegionValue manipulation operations. 

Yes.  What's that list?  Let's fill out IR.

It is an outstanding question in my mind if regions should be position-independent (or if you have to fix up pointers when you move/copy/serialize).  Right now they are position-independent but position-dependence is attractive for several reasons: it makes supporting supporting join and distinct operations easier, and it means we can just make the RegionValue a Long independent of the Region.  But for now, you should the Region that goes along with a RegionValue. (Right now the RegionValue class is a Region + Long.)